### PR TITLE
enhancement: revise feature list saving logic to reduce required api calls

### PR DESCRIPTION
**Pull request by gee-senbong**
## Description

This PR moves individual feature savings logic & feature list saving logic into an async task.

## Type of Change

Label this pull request to place it under the correct category in Release Notes:

|        **Pull Request Label**         | **Category in Release Notes** |
|:-------------------------------------:|:-----------------------------:|
|       `enhancement`, `feature`        |          🚀 Features          |
| `bug`, `refactoring`, `bugfix`, `fix` |    🔧 Fixes & Refactoring     |
|       `build`, `ci`, `testing`        |    📦 Build System & CI/CD    |
|              `breaking`               |      💥 Breaking Changes      |
|            `documentation`            |       📝 Documentation        |
|            `dependencies`             |    ⬆️ Dependencies updates    |

## Checklist

- [ ] I have read the [`CODE_OF_CONDUCT.md`](https://github.com/featurebyte/featurebyte/blob/main/CODE_OF_CONDUCT.md) and [`CONTRIBUTING.md`](https://github.com/featurebyte/featurebyte/blob/main/CONTRIBUTING.md) guides.
- [ ] I have written tests for the changes made.
- [ ] I have written docstrings in [NumpyDoc format](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard)
- [ ] I have labeled my Pull Request correctly

### CODE REVIEW

File: featurebyte/api/feature_list.py [hunk 1]
```diff
@@ -719,15 +693,40 @@ def save(
...
+        for feat in self.feature_objects.values():
+            feature_payloads.append(FeatureCreate(**feat.dict(by_alias=True)))
+
+        assert self.name is not None, "FeatureList name cannot be None"
+        feature_list_create = FeatureListCreateWithBatchFeatureCreation.create(
+            name=self.name,
+            features=feature_payloads,
+            conflict_resolution=conflict_resolution,
+            _id=self.id,
+        )
```
Comment (gee-senbong) [inline featurebyte/api/feature_list.py:708]: If we create a feature list of 204 saved features & deleted it, it takes 14.5s for this new feature list saving flow (current flow in the main branch takes only 12.5s).
The main reason is that it takes more time to (new flow)
(1) it takes more times to combine 204 feature graph into a feature cluster graph
(2) validates feature cluster graphs
than making 204 local api calls (main branch flow).

Comment (gee-senbong) [reply featurebyte/api/feature_list.py:708]: Combining list of feature graph into a feature clusters affects
(1) feature list document creation time (the most expensive part in create document creation 
https://github.com/featurebyte/featurebyte/blob/main/featurebyte/service/feature_list.py#L217-L223)
(2) the time taken to prepare batch feature cluster payload

File: featurebyte/worker/task/batch_feature_create.py [hunk 2]
```diff
@@ -149,5 +160,12 @@ async def execute(self) -> Any:
                 raise DocumentInconsistencyError("Inconsistent feature definition detected!")
 
             # update the progress
-            percent = 100 * (i + 1) / total
+            percent = start_percentage + (end_percentage - start_percentage) * (i + 1) / total
```
Comment (YS-L) [inline featurebyte/worker/task/batch_feature_create.py:163]: Another way is to pass in a ranged progress callback:

https://github.com/featurebyte/featurebyte/blob/a1f7dd966783d52f30fdebddfe9d60c67dc21649/featurebyte/common/progress.py#L9-L16

so this kind of arithmetic can be performed automatically.

**Review by YS-L — APPROVED** [review]

File: featurebyte/worker/task/batch_feature_create.py (hunk 2 above)
Comment (gee-senbong) [reply featurebyte/worker/task/batch_feature_create.py:163]: Will update this to use range process callback. 👍

File: featurebyte/worker/task/feature_list_batch_feature_create.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        feature_names = [feature.name for feature in payload.features]
+
+        # identify saved features
+        saved_feature_ids = set()
+        async for doc in self.app_container.feature_service.list_documents_iterator(
+            query_filter={"_id": {"$in": feature_ids}}
+        ):
+            saved_feature_ids.add(doc["_id"])
+
+        # construct conflict feature id to resolution feature id map
```
[the remark this answers is not in the document]
Comment (gee-senbong) [reply featurebyte/worker/task/feature_list_batch_feature_create.py:40]: Good catch, the logic here should be updated!

**Pull request closed without merging**